Marathon Provider support

gokmen opened the pull request:
cc/ @rjeczalik

gokmen committed:
ComputeUI: credential field required is optional, update label width

gokmen committed:
Provider: Marathon: add style for marathon provider

gokmen committed:
Provider: Marathon: add default stack template

gokmen committed:
Provider: Marathon: add config and enable provider as in beta mode

gokmen committed:
ComputeProvider: add Marathon provider

rjeczalik committed:
kites/config: fix klient.gz url

rjeczalik committed:
klient: rework entrypoint to use klient.gz from /mnt/mesos

rjeczalik committed:
provider/marathon: fetch klient.gz via marathon fetcher

rjeczalik committed:
klient: add TunnelID to force tunnelled connection

rjeczalik committed:
provider/marathon: disable healthcheck until reworked

rjeczalik committed:
klient: improve entrypoint

gokmen commented:
@rjeczalik I'm done with this PR waiting for your input, fyi.

rjeczalik committed:
provider/marathon: implement custom (*BaseStack).StateFunc method

rjeczalik committed:
provider/marathon: add debug support

rjeczalik committed:
klient/termial: improve logging

rjeczalik committed:
provider/marathon: fetch screen

rjeczalik committed:
klient/terminal: do not use sudo if not available

rjeczalik committed:
klient/entrypoint: init screen

rjeczalik committed:
provider/marathon: don't cache screen.tar.gz

rjeczalik committed:
klient/entrypoint: fix path to xterm-256color

rjeczalik commented:
@gokmen Assigning back. A number of changes, the most important one was to ship screen binary with klient. In order to have a single screen binary work across various linux distros I've compiled a static screen binary under Alpine Linux with musl replacement for glibc, with a small hack to workaround missing wtmp/utmp support in musl.

Since proxy does not work currently in development environment, I've tested webterm using my old evilkite client. I've also noticed that stack build modal sometimes losses events, especially when they are instant like in case of marathon.

Obligatory gif:

![pic](http://g.recordit.co/o3su09AweC.gif)

rjeczalik commented:
Can we also change default template to `cpus: 0.5` + `mem: 128`?

rjeczalik committed:
koding: reformat json files

rjeczalik committed:
provider/marathon: use unique string in app_id

rjeczalik committed:
klient: use websocket transport for tunnelserver connection

When multiple tunnels are connected at once, kite server
fails with:

  2016-11-14 12:02:50 [klient] ERROR    error sending: XHR session does not exist: eb8b4750f8f503a5fdbf
  2016-11-14T12:02:50.4Z   CRITICAL [tunnelclient][PID:30][logging/context.go:23] [transport] client fetch identifier error: sendError: XHR session does not exist: eb8b4750f8f503a5fdbf

rjeczalik commented:
@cihangir @ppknap Please review kloud- and klient-related code.

rjeczalik commented:
Rebased, split vendor update out to a separate PR: https://github.com/koding/koding/pull/9565

ppknap reviewed (approved):
LGTM

rjeczalik committed:
provider/marathon: add a test for unique labeling

rjeczalik commented:
Rebased and added a test for unique labeling: https://github.com/koding/koding/pull/9547/commits/c84ba8890fac913646764a6967245b1aa0f9beb7

rjeczalik committed:
klient/entrypoint: symlink /var/run/screen

cihangir commented on go/src/koding/kites/kloud/provider/marathon/testdata/multi-container.json.golden:
```diff
@@ -98,7 +111,7 @@
             {
               "command": {
-                "value": "curl -f -X GET http://$$HOST:$${PORT_56789}/kite"
+                "value": "true"
```
just a question, given the fact that i am not familiar with marathon,  why did we swicth from curl statement to true?

rjeczalik replied:
https://github.com/koding/koding/pull/9547/files#diff-12a6b9ca19c0ec2dc9242fd05655ce9eR26

The healtchecking is not needed to have working marathon provider - I wanted to be too correct to provide a healthcheck to ensure klient works. I need to both extend klient to allow for healthchecking (`curl .../kite` is not enough as it does not fail when klient has e.g. broken tunnel connection) and extend kloud provider for marathon with http healthchecks, but I'd need to map the ports for that.

rjeczalik replied:
I've changed it to not inject dummy health checks for now.

cihangir commented on go/src/koding/kites/tunnelproxy/client.go:
```diff
@@ -135,8 +136,18 @@ func NewClient(opts *ClientOptions) (*Client, error) {
 		optsCopy.Log = logging.NewCustom("tunnelclient", optsCopy.Debug)
 	}
 
+	// TODO(rjeczalik): fix production to use WebSocket by default
```
this is not viable for near future.

rjeczalik replied:
Yes I know, it uses websocket for tunnelserver only. This TODO has no deadline :-)

cihangir commented on go/src/koding/klient/terminal/command.go:
```diff
@@ -135,7 +135,11 @@ func screenSessions(username string) []string {
 	// visible from inside of container. Errors are ignored.
-	out, _ := exec.Command("ls", "/var/run/screen/S-"+username).Output()
+	out, err := exec.Command("ls", "/var/run/screen/S-"+username).Output()
+	if err != nil {
```
👍

cihangir commented on go/src/koding/klient/terminal/terminal.go:
```diff
@@ -253,7 +255,13 @@ func (t *Terminal) Connect(r *kite.Request) (interface{}, error) {
-	cmd := exec.Command("/usr/bin/sudo", args...)
+	var cmd *exec.Cmd
+
+	if _, err := exec.LookPath("sudo"); err == nil {
```
1) you might cache the sudo path
2) you could use the return value from LookPath `exec.Command("/usr/bin/sudo", args...)`

rjeczalik replied:
I will change it to os.Stat. I think the path must always be `/usr/bin/sudo`, executing it from different path can be a security threat I believe.

rjeczalik replied:
Done.

rjeczalik committed:
stack: require a credential for plan method

rjeczalik committed:
klient/terminal: change exec.LookPath with os.Stat

rjeczalik committed:
provider/marathon: pass cmd via env

This PR reworks entrypoint to execute cmd read from environment
variable. This fixes an issue, when global command was used for
all of the containers, causing only one entrypoint to be executed.
Instead the global command is removed and passed as env, so each
container executes a unique entrypoint, which then executes cmd.

rjeczalik committed:
provider/marathon: do not inject dummy health checks

rjeczalik commented:
I've added support for multiline `cmd`, so it's possible to run ubuntu:14.04 like:

```
provider:
  marathon:
    url: "${var.marathon_url}"
    basic_auth_user: "${var.marathon_basic_auth_user}"
    basic_auth_password: "${var.marathon_basic_auth_password}"

resource:
  marathon_app:
    # this is the name of your app
    app:
      # 
      cmd: |-
        apt-get update
        apt-get install -y ca-certificates
        /opt/kite/klient/klient
      
      container:
        docker:
        - image: ubuntu:14.04
          network: BRIDGE

      cpus: 1.0
      mem: 256
```

It does not work out of the box due to https://github.com/docker/docker/issues/3825.

rjeczalik committed:
provider/marathon: inject ca-certificates.crt into each container

rjeczalik commented:
Added one more change that injects ca-certificates.crt into container if it does not exist, ptal.

rjeczalik committed:
client: update default marathon template

rjeczalik committed:
klient: make klient default process when no cmd is specified

rjeczalik committed:
provider/marathon: use mongo state

rjeczalik commented:
@cihangir @gokmen ping

gokmen commented:
LGTM but I'm unable to approve it 

![image](https://cloud.githubusercontent.com/assets/42368/20494603/5c94b6f8-b026-11e6-98a3-99a6d07b629c.png)

rjeczalik commented:
@gokmen Did it for you! 💥
